compact.go

```diff
@@ -77,48 +83,58 @@ func (c Compact) Bytes() []byte {
 	return append([]byte{(topSixBits << 2) + 3}, b...)
 }
 
-func DecodeCompact(buffer *bytes.Buffer) (Compact, error) {
+func DecodeCompact[T BigNumbers](buffer *bytes.Buffer) (Compact[T], error) {
 	decoder := Decoder{Reader: buffer}
 	result := make([]byte, 16)
 	b, err := decoder.DecodeByte()
 	if err != nil {
-		return Compact{}, err
+		return Compact[T]{}, err
 	}
 	mode := b & 3
 	switch mode {
 	case 0:
-		return Compact(NewU128(big.NewInt(0).SetUint64(uint64(b >> 2)))), nil
+		value := NewU128(uint64(b >> 2))
+		result := interface{}(value).(T)
+		return Compact[T]{result}, nil
```

Comment:
You are always creating a `U128`. The idea was to create a `T` type with the decoded value.